### PR TITLE
🌱 Improve E2E ValidateFinalizers and ValidateOwnerRef

### PR DESCRIPTION
**Pull request by fabriziopandini**
**What this PR does / why we need it**:
This PR improves ValidateFinalizers and ValidateOwnerRef so the functions in input can get the namespaced name of the objects being processed.
This allows to handle differently e.g. secrets with CAPI certificate and secrets with cloud provider credentials

/area e2e-testing

### CODE REVIEW

File: test/framework/finalizers_helpers.go [hunk 1]
```diff
@@ -122,10 +124,11 @@ func getObjectsWithFinalizers(ctx context.Context, proxy ClusterProxy, namespace
 		Expect(err).ToNot(HaveOccurred())
 
 		setFinalizers := obj.GetFinalizers()
+		expectedFinalizers := allFinalizerAssertions[node.Object.Kind](types.NamespacedName{Namespace: node.Object.Namespace, Name: node.Object.Name})
```
Comment (chrischdi) [inline test/framework/finalizers_helpers.go:127]: There seems to be a nil pointer here

Comment (chrischdi) [reply test/framework/finalizers_helpers.go:127]:
```
In [It] at: /usr/local/go/src/runtime/panic.go:261 @ 05/28/24 15:03:26.623

runtime error: invalid memory address or nil pointer dereference

Full Stack Trace
  sigs.k8s.io/cluster-api/test/framework.getObjectsWithFinalizers({0x2b31788, 0xc0007afe50}, {0x2b43d30, 0xc000762360}, {0xc0008368a0, 0x12}, 0xc001681110, 0xc001702768)
  	/home/prow/go/src/sigs.k8s.io/cluster-api/test/framework/finalizers_helpers.go:127 +0x4cb
  sigs.k8s.io/cluster-api/test/framework.ValidateFinalizersResilience({0x2b31788, 0xc0007afe50}, {0x2b43d30, 0xc000762360}, {0xc0008368a0, 0x12}, {0xc000836918, 0x12}, 0xc001702768, {0xc0014151f0, ...})
  	/home/prow/go/src/sigs.k8s.io/cluster-api/test/framework/finalizers_helpers.go:75 +0xee
  sigs.k8s.io/cluster-api/test/e2e.init.func30.1.1({0x2b43d30, 0xc000762360}, {0xc0008368a0, 0x12}, {0xc000836918, 0x12})
  	/home/prow/go/src/sigs.k8s.io/cluster-api/test/e2e/quick_start_test.go:105 +0x353
  sigs.k8s.io/cluster-api/test/e2e.QuickStartSpec.func2.1()
  	/home/prow/go/src/sigs.k8s.io/cluster-api/test/e2e/quick_start.go:153 +0x46
  sigs.k8s.io/cluster-api/test/framework/clusterctl.ApplyCustomClusterTemplateAndWait({_, _}, {{0x2b43d30, 0xc000762360}, {0xc0009d0001, 0x74ef, 0x74f0}, {0xc000836918, 0x12}, {0xc0008368a0, ...}, ...}, ...)
  	/home/prow/go/src/sigs.k8s.io/cluster-api/test/framework/clusterctl/clusterctl_helpers.go:472 +0x163a
  sigs.k8s.io/cluster-api/test/framework/clusterctl.ApplyClusterTemplateAndWait({_, _}, {{0x2b43d30, 0xc000762360}, {{0xc000848bd0, 0x22}, {0xc00015a916, 0x31}, {0xc00015a948, 0x17}, ...}, ...}, ...)
  	/home/prow/go/src/sigs.k8s.io/cluster-api/test/framework/clusterctl/clusterctl_helpers.go:350 +0x9f1
  sigs.k8s.io/cluster-api/test/e2e.QuickStartSpec.func2()
  	/home/prow/go/src/sigs.k8s.io/cluster-api/test/e2e/quick_start.go:133 +0x748

```

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/10693/pull-cluster-api-e2e-blocking-main/1795467930796822528

**Comment by sbueringer** [comment]
In general lgtm so far

File: test/framework/finalizers_helpers.go (hunk 1 above)
Comment (fabriziopandini) [reply test/framework/finalizers_helpers.go]: Should be fixed now, but I have a follow up to investigate

File: test/e2e/quick_start_test.go [hunk 2]
```diff
@@ -29,7 +29,7 @@ import (
 	"sigs.k8s.io/cluster-api/test/framework/kubetest"
 )
 
-var _ = Describe("When following the Cluster API quick-start", func() {
+var _ = Describe("When following the Cluster API quick-start FP", func() {
```
Comment (sbueringer) [inline test/e2e/quick_start_test.go:32]:
```suggestion
var _ = Describe("When following the Cluster API quick-start", func() {
```

File: test/framework/finalizers_helpers.go [hunk 3]
```diff
@@ -125,7 +127,11 @@ func getObjectsWithFinalizers(ctx context.Context, proxy ClusterProxy, namespace
 
 		if len(setFinalizers) > 0 {
```
Comment (sbueringer) [inline test/framework/finalizers_helpers.go:128]: (I'm aware this was the same before)

This only validates finalizers if there are actually finalizers on the object, which means missing finalizers are accepted

Do I understand it correctly that basically we assume the finalizers are correct the first time we call getObjectsWithFinalizers and then in assertFinalizersExist we would detect if a finalizer is missing compared to the first time we retrieved them?

Comment (fabriziopandini) [reply test/framework/finalizers_helpers.go:143]: That's correct.
I have noticed this as well and now it should be fixed (we also check for missing finalizers + we check for additional finalizers that can pop up after we remove them)

Comment (sbueringer) [reply test/framework/finalizers_helpers.go:143]: Thx, nice one!

File: test/framework/finalizers_helpers.go [hunk 4]
```diff
@@ -134,24 +149,29 @@ func getObjectsWithFinalizers(ctx context.Context, proxy ClusterProxy, namespace
...
+					objKindNamespacedName, obj.GetFinalizers()))
 				continue
 			}
 
 			// verify if this resource has the appropriate Finalizers set
-			expectedFinalizers, assert := allFinalizerAssertions[obj.GetKind()]
+			expectedFinalizersF, assert := allFinalizerAssertions[obj.GetKind()]
 			if !assert {
+				// NOTE: this case should never happen because all the initialObjsWithFinalizers have been already checked
+				// against a finalizer assertion.
```
Comment (chrischdi) [inline test/framework/finalizers_helpers.go:170]: Shouldn't we then better Expect it to be true and fail instead of ignoring this error?

**Review by chrischdi — COMMENTED** [review]
One nit

/approve

/hold in case you want to change something at the comment

File: test/framework/finalizers_helpers.go (hunk 4 above)
Comment (sbueringer) [reply test/framework/finalizers_helpers.go:170]: Sounds reasonable to me

**Comment by chrischdi** [comment]
Either this or:

#10702 

Made e2e mink8s fail: https://testgrid.k8s.io/sig-cluster-lifecycle-cluster-api#capi-e2e-mink8s-main

**Comment by sbueringer** [comment]
`GetObjectVariableInto` is not used in core CAPI, so in any case this shouldn't lead to any trouble with e2e tests

**Pull request merged**